src/lib.rs

```diff
@@ -115,34 +115,45 @@ unsafe fn console_on_any(fds: &[DWORD]) -> bool {
 
 /// Returns true if there is an MSYS tty on the given handle.
 #[cfg(windows)]
-unsafe fn msys_tty_on(fd: DWORD) -> bool {
-    use std::{mem, slice};
-
+fn msys_tty_on(fd: DWORD) -> bool {
     use winapi::{
         ctypes::c_void,
         shared::minwindef::MAX_PATH,
         um::{
-            fileapi::FILE_NAME_INFO, minwinbase::FileNameInfo, processenv::GetStdHandle,
+            minwinbase::FileNameInfo, processenv::GetStdHandle,
             winbase::GetFileInformationByHandleEx,
         },
     };
 
-    let size = mem::size_of::<FILE_NAME_INFO>();
-    let mut name_info_bytes = vec![0u8; size + MAX_PATH * mem::size_of::<WCHAR>()];
-    let res = GetFileInformationByHandleEx(
-        GetStdHandle(fd),
-        FileNameInfo,
-        &mut *name_info_bytes as *mut _ as *mut c_void,
-        name_info_bytes.len() as u32,
-    );
+    /// Mirrors winapi::um::fileapi::FILE_NAME_INFO, giving it a fixed length that
+    /// we can stack allocate
+    #[repr(C)]
+    struct FILE_NAME_INFO {
+        FileNameLength: DWORD,
+        FileName: [WCHAR; MAX_PATH]
+    }
+    let mut name_info = FILE_NAME_INFO {
+        FileNameLength: 0,
+        FileName: [0; MAX_PATH]
+    };
+    let handle = unsafe {
+        // Safety: function has no invariants. an invalid handle id will cause
+        //         GetFileInformationByHandleEx to return an error
+        GetStdHandle(fd)
+    };
+    let res = unsafe {
+        // Safety: handle is valid, and buffer length is fixed
```

Comment:
**issue**: The handle is not necessarily valid, `GetStdHandle` can return an `INVALID_HANDLE_VALUE` ([ref docs](https://learn.microsoft.com/en-us/windows/console/getstdhandle)) and it's fragile to rely on this api not doing anything wrong when passed invalid data. I would be more comfortable with this error being checked in rust before calling the system api.

Reply:
`GetFileInformationByHandleEx` *should* be well-behaved here, as documented in the comments. Handling an invalid handle is one of its error states. That said, I'd never really argue against defensive programming especially around winapi!

Either way, I probably won't do anything more with PR without input from a maintainer